### ISSUE TITLE
Secure Webocket Closing Without any Response

### ISSUE THREAD

**Issue by Aismael** [Z-Wave-Me/home-automation#425]
I want to connect to an web-socket with an self signed certificate. The web-socket is on the /WebSocket Path of the Server. The server is tested with other clients.
First i post my code then I will describe my issues with the functionality and the Documentation of z-wave. 
```javascript 
var socket = new sockets.websocket("wss://192.168.10.108:443/WebSocket");
        socket.onclose = function (ev) {
            self.log("error", "closed", false);
            self.log("error", ev, true);
        };
        socket.onopen = function (ev) {
            self.log("error", "connected", false);
            self.log("error", ev, true);
            socket.send("ping");
        };
        socket.onmessage = function (ev) {
            self.log("error", "msg", false);
            self.log("error", ev, true);
        };
        socket.onerror = function (ev) {
            self.log("error", "error", false);
            self.log("error", ev, true);
        };
```
Issue 1:
The connection closing without an error or an status message,  and onclosed() gets called and onopen() not.
The documentation says on this point:    "Note that
if close() is called before connection was established, onclose() is not executed." So in my Opinion the connection has to be established. Why is it closing?

Issue 2: 
the documentation says:   "socket.websocket(url, [protocol]) ... Optional
protocol parameter can be used to specify protocol from server capabilities (comma
separated string), ...". 
My question, are there any Examples or Documentations for the optional parameters, and when not do anybody knowing sth. about this parameters.

sincerely Martin

**Comment by PoltoS** [comment]
>  "Note that if close() is called before connection was established, onclose() is not executed."
Looks this changed since and we missed this doc update.

Your connection is close because of the self-signed certificate. You can solve this by exporting environment variable SSL_TRUST_SELF_SIGNED=yes in the context where Z-Way runs. For example you can edit /etc/init.d/z-way-server for this.

You can also specify ssl_ca_filepath, ssl_cert_filepath, ssl_private_key_filepath like the following: 
new sockets.websocket(URL, null, ssl_ca_filepath, ssl_cert_filepath, ssl_private_key_filepath);

You certainly need only ssl_ca_filepath

**Comment by Aismael** [comment]
One more Question where is the EntryPoint of the ssl_ca_filepath? The Automation Folder or somewhere else? for Example 
```javascript
        var socket = new sockets.websocket("wss://192.168.10.108:443/WebSocket",null,"/opt/z-way-server/automation/modules/AropacControl/htdocs/Aeropac.cer");
```
or
```javascript
        var socket = new sockets.websocket("wss://192.168.10.108:443/WebSocket",null,"/modules/AropacControl/htdocs/Aeropac.cer");
```
Is not working

**Comment by Aismael** [comment]
And one other Question i have set the Enviroment Variable for Development testing and the Connection says
```json
 { "data": "Connection error" }
```

I tested the Websocket Server with "Smart Websocket Client" Chrome Extension, so it should be working.

**Comment by PoltoS** [comment]
The path for certs is absolute. It is passed to OpenSSL lib

**Comment by PoltoS** [comment]
Can you summarize your problems and provide how to reproduce them (if possible, with real public servers)

**Comment by Aismael** [comment]
Summarization of my Problems are

1. _**I Cant Connect to an Secure Websocket with an self Signed Certificate**_,
    even after edit /etc/init.d/z-way-server
   here is an Example for an Secure-Websocket-Server With the same Issue
   https://github.com/Aismael/Example-Websocket-server/
   it is an NodeJs Server can be started over    
   /nodejsWSSServer/node_modules/websockets/examples/server.js

2. The Sensor-Provider i want to connect with wss,
    make there own self Signed SSL Certificate after starting
    and each Provider has his own Certificate.
    So i have to Create an Module to connect to them 
    without an manipulated /etc/init.d/z-way-server.
    _**Is there some Way to do this

**Comment by PoltoS** [comment]
1. Please show how have you edited  /etc/init.d/z-way-server
strace -f -e trace=open,stat can help to see what happens during cert check
2. For security reasons you can no allow self signed in other way than by setting it in start script.
But you can specify your own CA to be used instead. Do your own CA and signd all certs using that CA

**Comment by Aismael** [comment]
I insert SSL_TRUST_SELF_SIGNED=yes in the file like you said before.

**Comment by PoltoS** [comment]
Can I see how? To make sure you did it correctly.

**Comment by Aismael** [comment]
Here is my Whole  /etc/init.d/z-way-server File
```sh
#! /bin/sh
### BEGIN INIT INFO
# Provides:               z-way-server
# Required-Start:
# Required-Stop:
# Default-Start:         2 3 4 5
# Default-Stop:   0 1 6
# Short-Description: RaZberry Z-Wave service
# Description:     Start Z-Way server for to allow Raspberry Pi talk with Z-Wave devices using RaZberry
### END INIT INFO

# Description: RaZberry Z-Way server
# Author: Yurkin Vitaliy <aivs@z-wave.me>

PATH=/bin:/usr/bin:/sbin:/usr/sbin
NAME=z-way-server
DAEMON_PATH=/opt/z-way-server
PIDFILE=/var/run/$NAME.pid
SSL_TRUST_SELF_SIGNED=yes
```

**Comment by PoltoS** [comment]
will not work. You need to either export this variable or add it right in the line of z-way-server start

I suggest to run Z-Way manually as:
SSL_TRUST_SELF_SIGNED=yes LD_LIBRARY_PATH=libs ./z-way-server

If it will work, check the script again

**Comment by Aismael** [comment]
you mean start the server in that way for example 

sudo SSL_TRUST_SELF_SIGNED=yes LD_LIBRARY_PATH=libs etc/init.d/z-way-server start
